File: src/main/java/ru/privatenull/pnlibrary/logging/MBox.java
package ru.privatenull.pnlibrary.logging;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

import ru.privatenull.pnlibrary.banner.PluginBanner;

/**
 * Строитель красивого многострочного отчёта для консоли.
 * Сохраняет порядок добавления компонентов и печатается методом {@link #show()}.
 */
public final class MBox {

    private final PluginLogger logger;
    private final String title;
    private final Map<String, PluginBanner.Entry> entries = new LinkedHashMap<>();
    private final Map<String, Throwable> errors = new LinkedHashMap<>();
    private boolean shown;

    MBox(PluginLogger logger, String title) {
        this.logger = Objects.requireNonNull(logger, "logger");
        this.title = PluginLogger.requireText(title, "title");
    }

    /** Добавляет успешно завершённый компонент. */
    public MBox ok(String component) {
        return ok(component, null);
    }

    /** Добавляет успешно завершённый компонент с пояснением. */
    public MBox ok(String component, String details) {
        return entry(component, PluginBanner.Status.OK, details);
    }

    /** Добавляет предупреждение. */
    public MBox warn(String component) {
        return warn(component, null);
    }

    /** Добавляет предупреждение с пояснением. */
    public MBox warn(String component, String details) {
        return entry(component, PluginBanner.Status.WARN, details);
    }

    /** Добавляет ошибку без исключения. */
    public MBox fail(String component) {
        return fail(component, (String) null);
    }

    /** Добавляет ошибку с понятным пояснением. */
    public MBox fail(String component, String details) {
        return entry(component, PluginBanner.Status.FAIL, details);
    }

    /**
     * Добавляет ошибку и запоминает исключение. При {@link #show()} рядом с
     * MBox будет напечатан полный stack trace.
     */
    public MBox fail(String component, Throwable error) {
        String name = PluginLogger.requireText(component, "component");
        Throwable checkedError = Objects.requireNonNull(error, "error");
        entry(name, PluginBanner.Status.FAIL, PluginLogger.throwableMessage(checkedError));
        errors.put(name, checkedError);
        return this;
    }

    /** Добавляет намеренно пропущенный компонент. */
    public MBox skip(String component) {
        return skip(component, null);
    }

    /** Добавляет намеренно пропущенный компонент с пояснением. */
    public MBox skip(String component, String details) {
        return entry(component, PluginBanner.Status.SKIP, details);
    }

    /**
     * Быстро показывает состояние опционального компонента.
     * Активный компонент получает {@code OK}, неактивный — {@code SKIP}.
     *
     * @param component название компонента
     * @param active запущен ли компонент
     * @return этот MBox
     */
    public MBox componentStatus(String component, boolean active) {
        return active
                ? ok(component, "Активен")
                : skip(component, "Не настроен");
    }

    /** Добавляет компонент с произвольным статусом. */
    public MBox entry(String component, PluginBanner.Status status, String details) {
        ensureNotShown();
        String name = PluginLogger.requireText(component, "component");
        entries.put(name, new PluginBanner.Entry(
                Objects.requireNonNull(status, "status"), details));
        if (status != PluginBanner.Status.FAIL) {
            errors.remove(name);
        }
        return this;
    }

    /**
     * Печатает блок. Один экземпляр MBox можно напечатать только один раз,
     * чтобы случайно не дублировать сообщения и stack trace.
     */
    public void show() {
        ensureNotShown();
        shown = true;
        logger.printBox(this);
    }

    /** Псевдоним {@link #show()} для кода, где удобнее название {@code log}. */
    public void log() {
        show();
    }

    /** @return заголовок блока */
    public String title() {
        return title;
    }

    /** @return неизменяемое представление добавленных компонентов */
    public Map<String, PluginBanner.Entry> entries() {
        return Collections.unmodifiableMap(entries);
    }

    Map<String, Throwable> errors() {
        return Collections.unmodifiableMap(errors);
    }

    PluginBanner.Status overallStatus() {
        if (contains(PluginBanner.Status.FAIL)) return PluginBanner.Status.FAIL;
        if (contains(PluginBanner.Status.WARN)) return PluginBanner.Status.WARN;
        if (!entries.isEmpty() && entries.values().stream()
                .allMatch(entry -> entry.status() == PluginBanner.Status.SKIP)) {
            return PluginBanner.Status.SKIP;
        }
        return PluginBanner.Status.OK;
    }

    private boolean contains(PluginBanner.Status status) {
        return entries.values().stream().anyMatch(entry -> entry.status() == status);
    }

    private void ensureNotShown() {
        if (shown) {
            throw new IllegalStateException("MBox has already been shown");
        }
    }
}

File: pnlibrary-bukkit/src/main/kotlin/ru/privatenull/pnlibrary/bukkit/logging/MBox.kt
package ru.privatenull.pnlibrary.bukkit.logging

import org.bukkit.ChatColor
import ru.privatenull.pnlibrary.bukkit.banner.PluginBanner
import java.util.Collections
import java.util.logging.Level

class MBox(
    private val logger: PluginLogger,
    val title: String,
) {
    private val _entries = linkedMapOf<String, PluginBanner.Entry>()
    private val _errors = linkedMapOf<String, Throwable>()
    private var shown = false

    val entries: Map<String, PluginBanner.Entry> get() = Collections.unmodifiableMap(_entries)

    fun ok(component: String, details: String? = null): MBox = entry(component, PluginBanner.Status.OK, details)
    fun warn(component: String, details: String? = null): MBox = entry(component, PluginBanner.Status.WARN, details)
    fun fail(component: String, details: String? = null): MBox = entry(component, PluginBanner.Status.FAIL, details)
    fun fail(component: String, error: Throwable): MBox {
        entry(component, PluginBanner.Status.FAIL, error.message ?: error.javaClass.simpleName)
        _errors[component] = error
        return this
    }
    fun skip(component: String, details: String? = null): MBox = entry(component, PluginBanner.Status.SKIP, details)

    fun componentStatus(component: String, active: Boolean): MBox =
        if (active) ok(component, "Активен") else skip(component, "Не настроен")

    fun entry(component: String, status: PluginBanner.Status, details: String? = null): MBox {
        check(!shown) { "MBox has already been shown" }
        _entries[component.trim()] = PluginBanner.Entry(status, details)
        if (status != PluginBanner.Status.FAIL) _errors.remove(component.trim())
        return this
    }

    fun show() {
        check(!shown) { "MBox has already been shown" }
        shown = true

        val console = logger.plugin.server.consoleSender
        val overallStatus = overallStatus()
        val color = overallStatus.color

        console.sendMessage("")
        console.sendMessage("${color}/\\_/\\")
        console.sendMessage("${color}( ${face(overallStatus)} )   ${ChatColor.AQUA}${logger.plugin.name}${ChatColor.DARK_GRAY} > ${ChatColor.WHITE}$title")
        console.sendMessage("${color}> ^ <")
        console.sendMessage("")

        for ((comp, entry) in _entries) {
            console.sendMessage("${ChatColor.DARK_GRAY}          > ${ChatColor.WHITE}$comp${ChatColor.DARK_GRAY}  ${entry.status.color}[ ${entry.status.displayName} ]")
            if (!entry.details.isNullOrBlank()) {
                console.sendMessage("${ChatColor.DARK_GRAY}            └ ${ChatColor.GRAY}${entry.details}")
            }
        }
        console.sendMessage("")

        for ((comp, err) in _errors) {
            logger.delegate.log(Level.SEVERE, "$title — $comp", err)
        }
    }

    fun log() = show()

    private fun overallStatus(): PluginBanner.Status {
        if (_entries.values.any { it.status == PluginBanner.Status.FAIL }) return PluginBanner.Status.FAIL
        if (_entries.values.any { it.status == PluginBanner.Status.WARN }) return PluginBanner.Status.WARN
        if (_entries.isNotEmpty() && _entries.values.all { it.status == PluginBanner.Status.SKIP }) return PluginBanner.Status.SKIP
        return PluginBanner.Status.OK
    }

    private fun face(status: PluginBanner.Status): String = when (status) {
        PluginBanner.Status.OK -> "^.^"
        PluginBanner.Status.WARN -> "o.o"
        PluginBanner.Status.FAIL -> "x.x"
        PluginBanner.Status.SKIP -> "-.-"
    }
}
